### ISSUE TITLE
[Typing] Rule enforcing type aliases in public interface must be public

### ISSUE THREAD

**Issue by randolf-scholz** [astral-sh/ruff#20916]
### Summary

It's good to see that so many libraries adopting type hints, but a recurring pattern I have observed is quite often, private type aliases are used to type hint public interfaces. This is a problem for users of such libraries because it means that if they want to extend functionality or wrap existing methods, they have to either:

1. import private members.
2. duplicate type aliases, which may not be stable between different versions of the library.

To avoid such issues happening by accident, I suggest a linting rule that checks that (top-level) type annotations in public interfaces are public.

### What it does

Checks that type annotations in public members only use public members (except type variables `TypeVar`, `TypeVarTuple` and `ParamSpec`)

### Example

```python
type _Alias = Literal["foo", "bar"]
type _PrivateAlias = str | int

class PublicClass:
    def plublic_method(self, arg: _Alias) -> None: ...
    def _private_method(self, arg: _PrivateAlias) -> None: ...
```

#### Use instead

```python
type PublicAlias = Literal["foo", "bar"]
type _PrivateAlias = str | int

class PublicClass:
    def plublic_method(self, arg: PublicAlias) -> None: ...
    def _private_method(self, arg: _PrivateAlias) -> None: ...
```

What is/should be OK is using private types in nested definitions (or maybe configurable with a `check_recursive` option)

```python
type _Foo = Literal["foo"]
type _Bar = Literal["bar"]
type PublicAlias = _Foo | _Bar  # OK

class PublicClass:
    def plublic_method(self, arg: PublicAlias) -> None: ...  # OK
```

**Comment by ntBre** [comment]
Thanks for the nice write-up! This reminds me of the [`private_interfaces`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_lint/builtin/static.PRIVATE_INTERFACES.html) warning in Rust.
